database/benchmark_database.go

```diff
@@ -0,0 +1,178 @@
+		for i := 0; i < b.N; i++ {
+			if err := db.Put(keys[i%count], values[i%count]); err != nil {
+				b.Fatalf("Unexpected error in Put %s", err)
+			}
+		}
+		b.ResetTimer()
+
+		// Deletes b.N values from the db
+		for i := 0; i < b.N; i++ {
+			if err := db.Delete(keys[i%count]); err != nil {
+				b.Fatalf("Unexpected error in Delete %s", err)
+			}
+		}
+	})
+}
+
+// BenchmarkBatchPut measures the time it takes to batch write.
+func BenchmarkBatchPut(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_write_batch", name, count, size), func(b *testing.B) {
+		batch := db.NewBatch()
+		if batch == nil {
+			b.Fatalf("db.NewBatch returned nil")
+		}
+		for i := 0; i < b.N; i++ {
+			if err := batch.Put(keys[i%count], values[i%count]); err != nil {
+				b.Fatalf("Unexpected error in db.Put: %s", err)
+			}
+			if err := batch.Write(); err != nil {
+				b.Fatalf("Unexpected error in batch.Write: %s", err)
+			}
+			batch.Reset()
+		}
+	})
+}
+
+// BenchmarkParallelGet measures the time it takes to read in parallel.
+func BenchmarkParallelGet(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_get_par", name, count, size), func(b *testing.B) {
+		b.RunParallel(func(pb *testing.PB) {
+			for i := 0; i < b.N; i++ {
+				if err := db.Put(keys[i%count], values[i%count]); err != nil {
+					b.Fatalf("Unexpected error in Put %s", err)
+				}
+			}
+			b.ResetTimer()
+
+			for i := 0; pb.Next(); i++ {
+				if _, err := db.Get(keys[i%count]); err != nil {
+					b.Fatalf("Unexpected error in Get %s", err)
+				}
+			}
+		})
+	})
+}
+
+// BenchmarkParallelPut measures the time it takes to write to the db in parallel.
+func BenchmarkParallelPut(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_write_par", name, count, size), func(b *testing.B) {
+		b.RunParallel(func(pb *testing.PB) {
+			// Write N values to the db
+			for i := 0; pb.Next(); i++ {
+				if err := db.Put(keys[i%count], values[i%count]); err != nil {
+					b.Fatalf("Unexpected error in Put %s", err)
+				}
+			}
+		})
+	})
+}
+
+// BenchmarkParallelDelete measures the time it takes to delete a (k, v) from the db.
+func BenchmarkParallelDelete(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_delete_par", name, count, size), func(b *testing.B) {
+		b.RunParallel(func(pb *testing.PB) {
+			// Writes random values of dbSize _dbSize_ to the database
+			for i := 0; pb.Next(); i++ {
+				if err := db.Put(keys[i%count], values[i%count]); err != nil {
+					b.Fatalf("Unexpected error in Put %s", err)
+				}
+			}
+			b.ResetTimer()
+
+			// Deletes b.N values from the db
+			for i := 0; pb.Next(); i++ {
```

Comment:
this part won't run. once pb.Next is exhausted it will always return false.
you can probably move Put outside of the RunParallel, or setup two RunParallel inside one benchmark to get the desired behavior.

also, ResetTimer will interfere with benchmark results, it spawns multiple goroutines, so you have to synchronize them if you want to get the correct results, otherwise some goroutines might be already doing actual work that is benchmarked when another goroutine calls b.ResetTimer